usecases/LDpred2_test/scripts/extended.sh

```diff
@@ -0,0 +1,57 @@
+LDP="$RSCRIPT $DIR_SCRIPTS/ldpred2.R --file-keep-snps $fileKeepSNPS \
+  --merge-by-rsid \
+  --col-stat beta --col-stat-se beta_se \
+  --col-snp-id rsid --col-chr chr --col-bp pos --col-A1 a1 --col-A2 a0 \
+  --out ${fileOut}_imputed.inf"
+
+# Create sumstats with characters in chromosome column
+echo "Test error: Character in chromosome column"
+row=$(tail -n 1  $fileInputSumStats)
+row=${row/,22,/,X,}
+fileSumstatsWithChar=$DIR_TESTS/data/sumstats_chr_char.txt
+head -n -1 $fileInputSumStats > $fileSumstatsWithChar
+echo "$row" >> $fileSumstatsWithChar
+dump=$( { $LDP --ldpred-mode inf --geno-file-rds $fileImputed.rds --sumstats $fileSumstatsWithChar; } 2>&1 )
+if [ $? -eq 1 ]; then echo "$dump"; exit; fi
+
+echo "Test error: Missing genotypes"
+dump=$( { $LDP --ldpred-mode inf --geno-file-rds $fileImpute.rds --sumstats $fileInputSumStats; } 2>&1 )
+if [ $? -eq 0 ]; then echo "No error received"; echo "$dump"; exit; fi
+echo "Test --geno-impute-zero"
+dump=$( { $LDP --ldpred-mode inf --geno-file-rds $fileImpute.rds --geno-impute-zero --sumstats $fileInputSumStats; } 2>&1 )
+if [ $? -eq 1 ]; then echo "$dump"; exit; fi
+echo "Test preimputed file"
+dump=$( { $LDP --ldpred-mode inf --geno-file-rds $fileImputed.rds --sumstats $fileInputSumStats; } 2>&1 )
+if [ $? -eq 1 ]; then echo "$dump"; exit; fi
+
+# Note that if files in --dir-genetic-maps do not exist, these will be downloaded. But I think error if the directory doesnt exist
+LDP="$RSCRIPT $DIR_SCRIPTS/ldpred2.R --file-keep-snps $fileKeepSNPS \
+  --merge-by-rsid \
+  --col-stat beta --col-stat-se beta_se \
+  --col-snp-id rsid --col-chr chr --col-bp pos --col-A1 a0 --col-A2 a1 \
+  --geno-file-rds $fileOutputSNPR --sumstats $fileInputSumStats"
+
+echo "TEST error: Bad ldpred mode"
+dump=$( { $LDP --ldpred-mode infinite --out $fileOut.inf; } 2>&1 )
+if [ $? -eq 0 ]; then echo "No error received"; exit; fi
+
+echo "TEST error: Bad imputation mode"
+dump=$( { $LDP --ldpred-mode inf --geno-impute bad-mode --out $fileOut.inf; } 2>&1 )
+if [ $? -eq 0 ]; then echo "No error received"; exit; fi
+
+# TEST: Complete runs of --ldpred-mode given by $LDPRED_MODES
+for MODE in $LDPRED_MODES; do
+ echo "Testing mode $MODE"
+ dump=$( { $LDP --ldpred-mode $MODE --out $fileOut.$MODE; } 2>&1 )
+ if [ $? -eq 1 ]; then echo "$dump"; exit; fi
+done
+echo "Test appending score to output file"
+export fileExisting=$fileOut.inf
+export scoreNameExisting=score # Exists from runs above
```

Comment:
@deepchocolate; I think the files `score` and `newScore` on the next line are not correctly defined. Either these should be explicitly defined (`score.txt`) or via some environment variable (`$score`). The `extended.R` script won't now load the correct files, causing the test crash.

Reply:
Not sure I follow... `$fileOut` is defined by `run.sh` to "public-data.score" which takes on a .inf or .auto extension in this place..

The only way I can replicate this error is by starting up Chrome otherwise it runs fine

Reply:
Hi. It's wasn't `$fileOut.inf` that I referred to. It was `score` and `newScore` on line 50-51. I now understand they are column names; however, the column name `newScore` is not found in the output file in the test. 
I added 
```
  print(c(fileExisting, scoreNameExisting, scoreNameNew))
  print(cnames)
  print(cnamesExp)
```
to the `test_that` function in `extended.R` script, yielding
```
[1] "/nrec/space/espenh/containers/usecases/LDpred2_test/output/public-data.score.inf"
[2] "score"                                                                           
[3] "newScore"                                                                        
[1] "FID"   "IID"   "score"
[1] "FID"      "IID"      "score"    "newScore"
```
so the output file is not merged as intended. 
`echo "$dump"` yielded
```
Test appending score to output file
Checking ability to merge score with file /nrec/space/espenh/containers/usecases/LDpred2_test/output/public-data.score.inf due to --out-merge
Loading backingfile: /nrec/space/espenh/containers/usecases/LDpred2_test/data/public-data3.rds 

### Reading LD reference meta-file from  /ldpred2_ref/map_hm3_plus.rds 

### Reading summary statistics /nrec/space/espenh/containers/usecases/LDpred2_test/data/tutorial_data/public-data3-sumstats.txt 
Loaded 50000 SNPs
Filtering SNPs using /REF/hapmap3/w_hm3.justrs 
Retained 49085 out of 50000 
Filtering SNPs based on --chr2use
Retained 49085 out of 49085 
Matching sumstats to genotypes
49,085 variants to be matched.
0 ambiguous SNPs have been removed.
45,009 variants have been matched; 22,591 were flipped and 30,040 were reversed.
Matching sumstats to LD reference
45,009 variants to be matched.
0 ambiguous SNPs have been removed.
44,553 variants have been matched; 0 were flipped and 0 were reversed.
Warning message:
In merge.data.table(as.data.table(sumstats4), as.data.table(info_snp),  :
  column names 'pos.ss' are duplicated in the result

### Loading LD reference from  /ldpred2_ref/ldref_hm3_plus/LD_with_blocks_chr@.rds 
         LD_with_blocks_chr1.rds : loading LD for 3495 out of 117534 SNPs
Runing unittests on output
```
I'm not sure if the Warning shown here is related.

What does Chrome (the browser?) have to do with this, I wonder?

Reply:
@espenhgn  As you can see in the dump 
```
### Loading LD reference from  /ldpred2_ref/ldref_hm3_plus/LD_with_blocks_chr@.rds 
         LD_with_blocks_chr1.rds : loading LD for 3495 out of 117534 SNPs
```
it stops at chromosome 1, so there is an error occuring there. If you have a similar issue as me, the error is not provided in the output (`$dump`) because the process is killed by the OS.
To see the error you have to replace this:
https://github.com/comorment/containers/blob/1bc3a659ef49681c6fe68492f823ae0dc3d2c359/usecases/LDpred2_test/scripts/extended.sh#L52

With `$LDP --ldpred-mode inf --name-score $scoreNameNew --out $fileExisting --out-merge`, i.e. don't assign the output to a variable.

I have no idea why chrome causes this error for me, but it must be something with the memory. I noted that Rstudio seem to cause the same problem. I think Rstudio uses QtWebEnginge (https://wiki.qt.io/QtWebEngine) which relies on chromium. So if you're runnning Rstudio while running the test I guess it could cause the same issue as for me

Reply:
> Hi. I'm not using Rstudio for this testing, just `bash` in a virtual machine setup. It has 8GB of RAM, which has also caused some memory issues with loading the reference files in some other example(s). I'm not sure if there are any alternatives to [`readRDS`](https://www.rdocumentation.org/packages/base/versions/3.6.2/topics/readRDS) that allow reading only a subset of the file in R, i.e., the elements `[ind.chr3, ind.chr3]`.

I don't use RStudio for testing either, but it's enough for me that Rstudio or Chrome is running for this problem to appear

Reply:
Ok, if you are confident this code runs with enough memory available, I'm fine with merging this as is of course. 
I suppose the chromium-based software like to buffer a lot of memory even if it is not actually used.

Reply:
Yes, I'm confident it works.. but I will test to see if some complete scoring runs can be dropped or maybe downsize the datasets used. I think the tests of the tutorial results are most important to run in full

Reply:
Ok, sounds reasonable. I just checked; simply reading in `/ldpred2_ref/ldref_hm3_plus/LD_with_blocks_chr1.rds` in R consumes about 1.9GB of system memory, so if there are options to "free" memory when data is no longer needed by the scripts it may help (`rm(foo); gc()` or something).

Reply:
Wow, I didn't realize it was that hungry! Good idea! I'll see if that can be implemented as well